src/mem/Pool.cc

```diff
@@ -77,11 +47,17 @@ MemPools::MemPools()
 MemImplementingAllocator *
 MemPools::create(const char *label, size_t obj_size)
 {
-    ++poolCount;
+    // TODO Use ref-counted Pointer for pool lifecycle management
+    // that is complicated by all the global static pool pointers.
+    // For now leak these Allocator descendants on shutdown.
+
+    MemImplementingAllocator *newPool;
     if (defaultIsChunked)
-        return new MemPoolChunked (label, obj_size);
+        newPool = new MemPoolChunked(label, obj_size);
     else
-        return new MemPoolMalloc (label, obj_size);
+        newPool = new MemPoolMalloc(label, obj_size);
+    pools.emplace_back(newPool);
```

Comment:
Several things here.

* With pools being "late binding" this **is** (indirectly) performance sensitive code. Any call to new for a previously unused object type may initiate pool creation during performance sensitive logic.

* raw-pointers do not have destructors called on scope exit - they "leak". So the implicit vs explicit destructor dangers do not exist here. These pool pointers are about to become ref-counted Pointer (followup PR) which implement move semantics to remain emplace-safe[1]. Also strict move safety is also why pool.back() gets returned instead of these newPool pointers - which may be nullptr (or any undefined state) after the emplace.

[1] FYI, the type of side effects emplace have on move is part of why move semantics require (MUST) that the moved-from object be left in a valid null state by explicit move operators/constructors.

Reply:
> With pools being "late binding" this is (indirectly) performance sensitive code. Any call to new for a previously unused object type may initiate pool creation during performance sensitive logic.

Code that quickly executes just a few times per Squid instance lifetime is not performance sensitive code. MemPools::create() is no exception to that rule.

> implicit vs explicit destructor dangers do not exist here

The dangers are related to the construction, not destruction. Those dangers are not bugs in the current code. They are bugs that may happen when this code is improved (e.g., when raw pointers become smart). Such bugs have happened to good developers improving very similar code (as detailed in my change request).

We should not be spending time on discussing complicated performance optimizations in some hypothetical future code in the context of a once-in-a-lifetime pool creation code. We should just use the safer push_back() call. In the unlikely event the actual future code needs a different solution, that different solution will be evaluated then, when all the variables are known.

Reply:
> Amos: move semantics require (MUST) that the moved-from object be left in a valid null state by explicit move operators/constructors.

There is no "valid null state" requirement for explicit move methods. If you disagree, please provide a reference.

There is a similar-sounding, but different [design choice](https://en.cppreference.com/w/cpp/utility/move) for STL objects: 

> STL: Unless otherwise specified, all standard library objects that have been moved from are placed in a "valid but unspecified state", meaning the object's class invariants hold (so functions without preconditions, such as the assignment operator, can be safely used on the object after it was moved from).

There is also a similar-sounding description of a [typical implementation](https://en.cppreference.com/w/cpp/language/move_constructor) of a move constructor:

> C++: Move constructors typically "steal" the resources held by the argument rather than make copies of them, and leave the argument in some valid but otherwise indeterminate state.

Same for move assignment operators.

P.S. This part of the discussion has nothing to do with the code in question and actual emplacement dangers. I am only trying to correct this particular misunderstanding because it is likely to bite future PRs that attempt to implement move optimizations in other contexts. Statistically speaking, our current expected probability of a move optimization-related proposal being bug-free is close to zero.

Reply:
> > With pools being "late binding" this is (indirectly) performance sensitive code. Any call to new for a previously unused object type may initiate pool creation during performance sensitive logic.
> 
> Code that quickly executes just a few times per Squid instance lifetime is not performance sensitive code. MemPools::create() is no exception to that rule.
> 

FWIW, that "rule" seems to be vague and does not have reliable scope for what performance impact "quickly" and "just a few times" are having.

FYI, The definition I use for performance sensitive (code operating during action X, ie. from point A to point B in code) can be deterministically defined in terms of which specific lines of code are relevant. Allowing policy defined groups of actions with different amount of acceptable impact or size/speed optimization targets.


> > implicit vs explicit destructor dangers do not exist here
> 
> The dangers are related to the construction, not destruction.

Then I have completely missed what your point was. The references you provided as examples are all about bugs caused by code calling destructors twice (on moved and moved-from objects) after emplace() calls a badly designed move-constructor.
 These are expected "undefined behaviour" effects of buggy class X being passed around. Not dangers of using emplace general. eg. the same bugs would occur if instances of such class X are returned by-value from any custom code.


> > Amos: move semantics require (MUST) that the moved-from object be left in a valid null state by explicit move operators/constructors.
> 
> There is no "valid null state" requirement for explicit move methods. If you disagree, please provide a reference.
> 

C++11 document N337 section 12.2.1 on xvalue (temporary object) states.
"
Even when the creation of the temporary object is unevaluated (Clause 5) or otherwise avoided (12.8 [copy/move construction]), all the semantic restrictions shall be respected as if the temporary object had been created and later destroyed.
"

The relevant "semantic restrictions" for emplace_back() being the MoveConstructible trait defined in section 17.6.3.1 table 20:
"
 Note:rv must still meet the requirements of the library component that is using it. The operations listed in those requirements must work as specified whether rv has been moved from or not.
"
 [ ```rv``` being the moved-from object]

Thus the value "shall be" left valid as per default-initialization (ergo: nullptr).

Reply:
> > > Amos: With pools being "late binding" this is (indirectly) performance sensitive code. Any call to new for a previously unused object type may initiate pool creation during performance sensitive logic.
 
> > Alex: Code that quickly executes just a few times per Squid instance lifetime is not performance sensitive code. MemPools::create() is no exception to that rule.
 
> Amos: FWIW, that "rule" seems to be vague and does not have reliable scope for what performance impact "quickly" and "just a few times" are having.

Agreed. Fortunately, any reasonable definition of "quickly" and "few" would exclude MemPools::create() from being classified as performance-sensitive in this Squid-specific context, and that is all that matters as far as this change request is concerned. In other words, the proposed rule of thumb is good enough to easily make the correct decision here (and in the vast majority of real cases). That is what rules of thumb are best for -- quickly and correctly handling the vast majority of use cases.

 
> FYI, The definition I use for performance sensitive (code operating during action X, ie. from point A to point B in code) can be deterministically defined in terms of which specific lines of code are relevant. Allowing policy defined groups of actions with different amount of acceptable impact or size/speed optimization targets.

I do not know what definition of "performance sensitive" you are using. If you would like to establish some formal terminology for future shared use, please make a specific proposal. FWIW, I am sure that any successful outcome of that activity will _not_ classify MemPools::create() (in the current or any foreseeable Squid code context) as "performance sensitive", but I am prepared to be proven wrong.

> > > implicit vs explicit destructor dangers do not exist here

> > The dangers are related to the construction, not destruction.

> The references you provided as examples are all about bugs caused by code calling destructors twice (on moved and moved-from objects)

Not all of them. Yes, one of the provided problems results in double-destruction. There are also problems that have nothing to do with destructors (e.g., look for `1<<20` in https://abseil.io/tips/112). And even if we focus just on that double-destruction problem on [SO](https://stackoverflow.com/a/36919571), its _cause_  was in _construction_ (and that is where the emplace() danger is):

> after emplace() calls a badly designed move-constructor.

In the SO example, emplace() does not call a move constructor. Moreover, I would not call std::unique_ptr move constructor "badly designed". In fact, no move constructor can be badly designed -- they all have pretty much the same design (dictated by C++ language rules). Only their implementations differ.

In the SO example, emplace() calls an explicit `std::unique_ptr<T>::unique_ptr(T*)` constructor that converts a raw T pointer into a unique pointer. There is nothing wrong with that constructor either -- it is essential and well-designed.

The problems here are not triggered by any specific constructor. They are triggered by emplace_back() secret use of a  type conversion that is supposed to be explicit (via the use of an explicit constructor in the method implementation invisible to the code writer). Implicit type conversions can be very dangerous, of course!

In other words, emplace_back() essentially makes an explicit constructor implicit, effectively removing protections that explicit constructors are supposed to provide, such as exposing dangerous type conversions to the developer (who normally has to type the constructor name explicitly to get that conversion).

> These are expected "undefined behaviour" effects of buggy class X being passed around. 

There are no buggy classes in that SO problem. All classes are OK! Emplace() screws up code where there are no class-level bugs: `emplace_back(a)` compiles where the code author would really want it to fail (and where `push_back(a)` does fail to compile, exposing the bug in the caller).

----

> > > Amos: move semantics require (MUST) that the moved-from object be left in a valid null state by explicit move operators/constructors.

> > There is no "valid null state" requirement for explicit move methods. If you disagree, please provide a reference.
 
> C++11 document N337 section 12.2.1 on xvalue (temporary object) states. " Even when the creation of the temporary object is unevaluated (Clause 5) or otherwise avoided (12.8 [copy/move construction]), all the semantic restrictions shall be respected as if the temporary object had been created and later destroyed. "

That text places no restrictions on the move constructor! It places restrictions on what happens when the move constructor call is _avoided_. In simple in-scope terms, it tells the compiler, "If you were supposed to create a temporary object X (e.g., to pass to a move constructor), but you decided not to create X (and not call that move constructor), then you shall still respect all the semantic restrictions that would have existed if X had been created and later destroyed". 

For example, that temporary object may be avoided by applying move elision optimization. In fact, the C++11 standard gives that example right below the 12.2.1 text you have sited (at least it does in N3242 copy that I could [find](https://stackoverflow.com/a/7747129)).

> The relevant "semantic restrictions" for emplace_back() being the MoveConstructible trait defined in section 17.6.3.1 table 20: " Note:rv must still meet the requirements of the library component that is using it. The operations listed in those requirements must work as specified whether rv has been moved from or not. " [ `rv` being the moved-from object]

std::list::emplace_back() does not place MoveConstructible restrictions (on anything). It places [EmplaceConstructible](https://en.cppreference.com/w/cpp/named_req/EmplaceConstructible) restrictions on std::list value_type. In simpler terms, we can use emplace_back(x) when the list allocator can create value_type object in-place, using supplied parameter `x`. This has nothing to do with semantic restrictions on the move constructor.
 
> Thus the value "shall be" left valid as per default-initialization (ergo: nullptr).

The MoveConstructible requirement does not mention default-initialization. Furthermore, MoveConstructible explicitly places no restrictions on the `rv` value after construction. Here are citations from two sources:

* N3242: `rv` remains a valid object. Its state is unspecified.
* [cppreference](https://en.cppreference.com/w/cpp/named_req/MoveConstructible): The new value of `rv` is unspecified.

The only restriction is that `rv` must be a valid object (e.g., we can assign a new value to it without fear). I do not understand how one can get from MoveConstructible to "default-initialization".

In summary, neither the inapplicable specs nor the applicable ones place a "valid null state" or even "valid default-initializated state" requirement on explicit move methods. The actual post-move requirements are very simple and allow for various performance optimizations -- any valid state would do!